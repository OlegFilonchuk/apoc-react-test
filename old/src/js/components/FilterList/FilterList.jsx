import React from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';
import classNames from 'classnames';

import './FilterList.less';

export default class FilterList extends React.Component {
  static BASE_CLASS = 'filterlist';

  shouldComponentUpdate(nextProps) {
    return !isEqual(nextProps, this.props);
  }

  get buildOptions() {
    return this.props.options.map(({ name, value }) => (
      <li key={name} className={`${FilterList.BASE_CLASS}__option`}>
        <span>{name}</span>
        <button
          className={`${FilterList.BASE_CLASS}__option-remove sc-icon-filled-circle-close sc-txt-primary`}
          onClick={() => this.remove(value)}
        />
      </li>
    ));
  }

  get finalClassName() {
    return classNames(FilterList.BASE_CLASS, this.props.className);
  }

  remove = value => {
    this.props.onFilterRemove(value);
    this.props.onChange(this.props.options.filter(option => option.value.name !== value.name));
  };

  removeAll = () => {
    this.props.onChange([]);
    this.props.onClearAll();
  };

  render() {
    return (
      <ul className={this.finalClassName}>
        {this.buildOptions}
        <li className={`${FilterList.BASE_CLASS}__remove-all`}>
          <button onClick={this.removeAll} className="sc-txt-primary">
            Clear All
          </button>
        </li>
      </ul>
    );
  }
}

FilterList.propTypes = {
  options: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.object])
    })
  ),
  onChange: PropTypes.func,

  /**
   * Any custom class names
   */
  className: PropTypes.string,

  /**
   * For backward compatibility we are leaving `onChange`
   * but it has major drawback: giving you only list of items which left after change.
   *
   * Callbacks above are giving more granular control and they are REPLACEMENT for `onChange`
   */
  onClearAll: PropTypes.func,
  onFilterRemove: PropTypes.func
};

FilterList.defaultProps = {
  options: [],
  onChange: () => true,
  onClearAll: () => true,
  onFilterRemove: () => true,
  className: ''
};
