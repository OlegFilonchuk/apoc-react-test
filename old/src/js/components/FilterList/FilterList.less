@import '~chameleon/less/variables.less';

@filterlist-vertical-spacing-scale: 2;
@filterlist-horizontal-spacing-scale: 1.5;

@padding-v: (@padding-base-vertical / @filterlist-vertical-spacing-scale);
@padding-h: (@padding-base-horizontal / @filterlist-horizontal-spacing-scale);

.filterlist {
  padding: 0;
  margin: 0;
  list-style-type: none;
  display: inline-block;

  &__option,
  &__remove-all {
    display: inline-block;
    margin: 0 0 0 (@margin-btn / @filterlist-horizontal-spacing-scale);

    button {
      background: transparent;
      border: none;

      &:hover {
        color: @primary-hover;
      }

      &:focus {
        outline: none;
      }
    }
  }

  &__option {
    background-color: #eaeaea;
    padding: @padding-v 0 @padding-v @padding-h;

    button {
      font-size: 110%;
      padding: 0 @padding-h;
    }

    &--dummy {
      opacity: 0;
      width: 0;
      padding-left: 0;
      padding-right: 0;
      margin-left: 0;
      margin-top: 0;
    }
  }

  &__remove-all {
    background-color: transparent;
    color: @primary;
  }
}
