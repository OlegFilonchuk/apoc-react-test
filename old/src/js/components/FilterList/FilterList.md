The component is intended to be used as controlled component. It does not keep its own state.

## FilterList static example

    const options = [
      {name: 'Business Systems', value: 'value1'},
      {name: 'iPhone', value: 'value2'},
      {name: 'Imported Scenarios', value: 'value3'},
      {name: 'Social Networking', value: 'value4'}
    ];

    <div>
        Filtering by: <FilterList options={options} />
    </div>

## FilterList editable example

    const options = [
      {name: 'Business Systems', value: 'value1'},
      {name: 'iPhone', value: 'value2'},
      {name: 'Imported Scenarios', value: 'value3'},
      {name: 'Social Networking', value: 'value4'}
    ];
    const resetOptions = () => {
      setState({options: [...options]});
    };
    const onChange = (options) => {
      setState({options: [...options]})
    };
    initialState = {options: [...options]};

    <div>
        <Button onClick={resetOptions} className="sc-btn-primary">restore options</Button><br /><br />
        Filtering by: <FilterList options={state.options} onChange={onChange} />
    </div>