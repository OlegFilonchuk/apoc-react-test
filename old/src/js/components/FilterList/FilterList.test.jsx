import React from 'react';
import { mount } from 'enzyme';

import FilterList from './FilterList';

const defaultOptions = [
  { name: 'Business Systems', value: 'value1' },
  { name: 'iPhone', value: 'value2' },
  { name: 'Imported Scenarios', value: 'value3' },
  { name: 'Social Networking', value: 'value4' }
];

describe('<FilterList />', () => {
  it('should render <FilterList />', () => {
    const component = mount(<FilterList options={defaultOptions} />);

    expect(component).toMatchSnapshot();
  });

  describe('when remove is clicked should call onChange', () => {
    const onChange = jest.fn();
    const component = mount(<FilterList options={defaultOptions} onChange={onChange} />);
    const itemToRemove = 2;

    component
      .find('button')
      .at(itemToRemove)
      .simulate('click');

    it('atleast once', () => {
      expect(onChange).toBeCalled();
    });

    it('with the right argument', () => {
      expect(onChange).toBeCalledWith(
        [...defaultOptions].filter(option => option.value.name !== defaultOptions[itemToRemove].value.name)
      );
    });
  });

  describe('when remove all is clicked should call onChange', () => {
    const onChange = jest.fn();
    const component = mount(<FilterList options={defaultOptions} onChange={onChange} />);

    component.find('.filterlist__remove-all button').simulate('click');

    it('atleast once', () => {
      expect(onChange).toBeCalled();
    });

    it('with empty array', () => {
      expect(onChange).toBeCalledWith([]);
    });
  });
});
