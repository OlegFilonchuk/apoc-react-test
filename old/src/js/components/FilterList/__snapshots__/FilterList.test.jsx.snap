// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterList /> should render <FilterList /> 1`] = `
<FilterList
  className=""
  onChange={[Function]}
  onClearAll={[Function]}
  onFilterRemove={[Function]}
  options={
    Array [
      Object {
        "name": "Business Systems",
        "value": "value1",
      },
      Object {
        "name": "iPhone",
        "value": "value2",
      },
      Object {
        "name": "Imported Scenarios",
        "value": "value3",
      },
      Object {
        "name": "Social Networking",
        "value": "value4",
      },
    ]
  }
>
  <ul
    className="filterlist"
  >
    <li
      className="filterlist__option"
      key="Business Systems"
    >
      <span>
        Business Systems
      </span>
      <button
        className="filterlist__option-remove sc-icon-filled-circle-close sc-txt-primary"
        onClick={[Function]}
      />
    </li>
    <li
      className="filterlist__option"
      key="iPhone"
    >
      <span>
        iPhone
      </span>
      <button
        className="filterlist__option-remove sc-icon-filled-circle-close sc-txt-primary"
        onClick={[Function]}
      />
    </li>
    <li
      className="filterlist__option"
      key="Imported Scenarios"
    >
      <span>
        Imported Scenarios
      </span>
      <button
        className="filterlist__option-remove sc-icon-filled-circle-close sc-txt-primary"
        onClick={[Function]}
      />
    </li>
    <li
      className="filterlist__option"
      key="Social Networking"
    >
      <span>
        Social Networking
      </span>
      <button
        className="filterlist__option-remove sc-icon-filled-circle-close sc-txt-primary"
        onClick={[Function]}
      />
    </li>
    <li
      className="filterlist__remove-all"
    >
      <button
        className="sc-txt-primary"
        onClick={[Function]}
      >
        Clear All
      </button>
    </li>
  </ul>
</FilterList>
`;
